include/dali/core/tensor_shape.h

```diff
@@ -1473,6 +1473,48 @@ TensorListShape<output_ndim> sample_range(const TensorListShape<ndim> &in,
   return ret;
 }
 
+/**
+ * @brief Unfolds outer dimension of TensorShape in TensorShapeList
+ * unfold_outer_dim([[2, 1], [1, 3]]) -> [1, 1, 3]
+ */
+template <int ndim>
+auto unfold_outer_dim(const TensorListShape<ndim> &shapes)
+    -> TensorListShape<ndim != DynamicDimensions ? ndim - 1 : DynamicDimensions> {
+  static_assert(ndim == DynamicDimensions || ndim >= 1,
+                "Can not unfold TensorShapeList of shapes with shape dimension 0");
+  DALI_ENFORCE(shapes.sample_dim() >= 1,
+               "Can not unfold TensorShapeList of shapes with shape dimension 0");
+
+  constexpr static int sout_dim = ndim != DynamicDimensions ? ndim - 1 : DynamicDimensions;
+  const int rout_dim = shapes.sample_dim() - 1;
+
+  int nshapes = 0;
+
+  for (int i = 0; i < shapes.size(); ++i) {
+    nshapes += shapes[i][0];
+  }
+
+  TensorListShape<sout_dim> result{nshapes, rout_dim};
+
+  if (rout_dim == 0) {
+    // NB: we don't need to create any shapes of dim equal 0
+    return result;
+  }
+
+  int64_t nouter_dim;
+
+  for (int i = 0, snum = 0; i < shapes.size(); ++i, snum += nouter_dim) {
+    auto shape = shapes[i];
+    nouter_dim = shape[0];
+    auto subshape = shape.last(rout_dim);
+    for (int j = 0; j < nouter_dim; j++) {
+      result.set_tensor_shape(snum + j, TensorShape<sout_dim>(subshape));
+    }
```

Comment:
This is technically correct, but for DynamicDimensions it's quite involved (in terms of what needs to happen - SmallVector isn't very heavy, but there are even lighter options. I'd recommend using `tensor_shape_span`.
```suggestion
    auto shape = shapes.tensor_shape_span(i);
    auto subshape = make_span(shape.begin() + 1, shape.end());
    int nouter_dim = shape[0];
    for (int j = 0; j < nouter_dim; j++) {
      result.set_tensor_shape(snum + j, subshape);
    }
```
Notes:
* there's a bit of a mess regarding sample indices - they should generally be `int`, we don't consider batches longer than INT_MAX. I know that some functions take it as int64_t, but it should be fixed at some point
* `span` is just a pointer + length - we don't need to copy the shape of each sample multiple times
* `set_tensor_shape` doesn't need a `TensorShape` - it can work with an arbitrary collection, including `span`

Reply:
Done